### PR TITLE
feat(TldrawUi): allow user override part of UI by passing uiComponents

### PR DESCRIPTION
**Pull request by Rokixy**
Add `<UiComponentsProvider />` and `useUiComponents()`. User could pass `uiComponents` as a prop of `<Tldraw />` or `<TldrawUi />`.

### Change Type

- [ ] `patch` — Bug fix
- [x] `minor` — New feature
- [ ] `major` — Breaking change
- [ ] `dependencies` — Changes to package dependencies[^1]
- [ ] `documentation` — Changes to the documentation only[^2]
- [ ] `tests` — Changes to any test code only[^2]
- [ ] `internal` — Any other changes that don't affect the published package[^2]
- [ ] I don't know

[^1]: publishes a `patch` release, for devDependencies use `internal`
[^2]: will not publish a new version

### Release Notes

- Add `uiComponents` as a prop of `<Tldraw />` or `<TldrawUi />` for overriding specific parts of UI.

### Relative Issue
https://github.com/tldraw/tldraw/issues/2677

### CODE REVIEW

**Comment by Rokixy** [comment]
Didn't mean to delete the `api.json` files. How can I get them back?

**Comment by Rokixy** [comment]
I'd appreciate it if you could review this PR for me. @steveruizok @TodePond

**Comment by steveruizok** [comment]
Thanks for the PR! I'll take a look. We've wanted to have a more composable / configurable UI for tldraw, so interested in seeing your approach.

File: packages/tldraw/src/lib/ui/hooks/useUiComponents.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+					ExitPenMode: DefaultExitPenMode,
+					StopFollowing: DefaultStopFollowing,
+					StylePanel: DefaultStylePanel,
+					ToastViewport: DefaultToastViewport,
+					Toasts: DefaultToasts,
+					Toolbar: DefaultToolbar,
+					Button: DefaultButton,
+					...uiComponents,
+				}),
+				[uiComponents]
```
Comment (judicaelandria) [inline packages/tldraw/src/lib/ui/hooks/useUiComponents.tsx:68]: love your approach on this, i just have a suggestion just to follow the code style in the repo and let's type it
```tsx
  const components: TLUiComponent & BaseUiComponents= useMemo(() => ({
                                          BackToContent: DefaultBackToContent,
					DebugPanel: DefaultDebugPanel,
					Dialogs: DefaultDialogs,
					FollowingIndicator: DefaultFollowingIndicator,
					HelpMenu: DefaultHelpMenu,
					MenuZone: DefaultMenuZone,
					NavigationZone: DefaultNavigationZone,
					ExitPenMode: DefaultExitPenMode,
					StopFollowing: DefaultStopFollowing,
					StylePanel: DefaultStylePanel,
					ToastViewport: DefaultToastViewport,
					Toasts: DefaultToasts,
					Toolbar: DefaultToolbar,
					Button: DefaultButton,
					...uiComponents,
  }),[uiComponents])
  ...
```

**Comment by steveruizok** [comment]
This PR should show what I meant in terms of API design: https://github.com/tldraw/tldraw/pull/2796

**Comment by Rokixy** [comment]
> This PR should show what I meant in terms of API design: #2796 

I'm glad to see similar content has already been merged in #2796 . Perhaps we could close this PR now?

**Pull request closed without merging**